File: tests/java-engine/EngineTest.java
import com.smartpantry.data.SeedData;
import com.smartpantry.logic.*;
import java.time.LocalDate;
import java.util.*;

public class EngineTest {
    static int pass = 0, fail = 0;
    static void t(String n, boolean c) { if (c) pass++; else { fail++; System.out.println("FAIL " + n); } }

    static class S implements Stock {
        long id; String name, unit, exp; double q;
        S(long id, String name, double q, String unit, String exp) { this.id = id; this.name = name; this.q = q; this.unit = unit; this.exp = exp; }
        public long getId() { return id; } public String getName() { return name; } public double getQuantity() { return q; }
        public String getUnit() { return unit; } public String getExpiryDate() { return exp; }
    }
    static class Rq implements Requirement {
        String n, u; double q; Rq(SeedData.Ing i) { n = i.name; q = i.quantity; u = i.unit; }
        Rq(String n, double q, String u) { this.n = n; this.q = q; this.u = u; }
        public String getName() { return n; } public double getQuantity() { return q; } public String getUnit() { return u; }
    }
    static List<Rq> reqs(SeedData.SeedRecipe r) { List<Rq> l = new ArrayList<>(); for (SeedData.Ing i : r.ingredients) l.add(new Rq(i)); return l; }

    public static void main(String[] a) {
        LocalDate today = LocalDate.of(2026, 9, 20);
        t("tomatoes", IngredientNormalizer.normalize("Tomatoes").equals("tomato"));
        t("potatoes", IngredientNormalizer.normalize("potatoes").equals("potato"));
        t("fresh large", IngredientNormalizer.normalize("Fresh large tomatoes").equals("tomato"));
        t("spaghetti", IngredientNormalizer.normalize("Spaghetti").equals("pasta"));
        t("evoo", IngredientNormalizer.normalize("extra virgin olive oil").equals("olive oil"));
        t("cookie", IngredientNormalizer.normalize("cookie").equals(IngredientNormalizer.normalize("cookies")));
        t("hummus", IngredientNormalizer.normalize("hummus").equals("hummus"));
        t("parens", IngredientNormalizer.normalize("Tomatoes (cherry)").equals("tomato"));
        t("chilli", IngredientNormalizer.normalize("Chilli flakes").equals(IngredientNormalizer.normalize("chili flake")));

        List<S> pantry = new ArrayList<>(); long id = 1;
        for (SeedData.DemoItem d : SeedData.DEMO_PANTRY)
            pantry.add(new S(id++, d.name, d.quantity, d.unit, d.expiryOffsetDays == null ? null : today.plusDays(d.expiryOffsetDays).toString()));
        StringBuilder ready = new StringBuilder(), almost = new StringBuilder();
        int rid = 1;
        for (SeedData.SeedRecipe r : SeedData.RECIPES) {
            MatchResult m = RecipeMatcher.evaluate(rid, reqs(r), pantry, true, today, 3);
            if (m.isReady()) ready.append(ready.length() > 0 ? "," : "").append(rid);
            if (m.isAlmostThere()) almost.append(almost.length() > 0 ? "," : "").append(rid);
            rid++;
        }
        System.out.println("READY " + ready + " | ALMOST " + almost);
        t("ready set", ready.toString().equals("1,2,5,6,8,11,13")); t("almost set", almost.toString().equals("3,4,15"));
        t("20 recipes", SeedData.RECIPES.size() == 20);

        // remove cheddar -> 1, 11, 13 vanish
        List<S> noCheese = new ArrayList<>(pantry); noCheese.removeIf(s -> s.name.equals("Cheddar cheese"));
        t("cheddar removal", !RecipeMatcher.evaluate(1, reqs(SeedData.RECIPES.get(0)), noCheese, true, today, 3).isReady()
                && !RecipeMatcher.evaluate(11, reqs(SeedData.RECIPES.get(10)), noCheese, true, today, 3).isReady()
                && RecipeMatcher.evaluate(2, reqs(SeedData.RECIPES.get(1)), noCheese, true, today, 3).isReady());
        // exact quantity boundary
        SeedData.SeedRecipe chick = SeedData.RECIPES.get(4);
        for (double[] c : new double[][]{{300, 1}, {299, 0}}) {
            List<S> p = new ArrayList<>(Arrays.asList(new S(1, "chicken breast", c[0], "g", null), new S(2, "rice", 200, "g", null),
                    new S(3, "onion", 1, "pcs", null), new S(4, "garlic", 2, "pcs", null), new S(5, "soy sauce", 2, "tbsp", null)));
            t("boundary " + c[0], RecipeMatcher.evaluate(5, reqs(chick), p, true, today, 3).isReady() == (c[1] == 1));
        }
        // unit differences
        for (Object[] c : new Object[][]{{30.0, "ml", true}, {29.0, "ml", false}, {0.25, "cup", true}, {1.0, "l", true}}) {
            List<S> p = new ArrayList<>(Arrays.asList(new S(1, "Chicken", 150, "g", null), new S(6, "chicken breast", 150, "g", null), new S(2, "rice", 0.2, "kg", null),
                    new S(3, "onions", 1, "pcs", null), new S(4, "garlic", 2, "pcs", null), new S(5, "soy sauce", (Double) c[0], (String) c[1], null)));
            t("unit " + c[0] + c[1], RecipeMatcher.evaluate(5, reqs(chick), p, true, today, 3).isReady() == (Boolean) c[2]);
        }
        // mass covers count
        List<Rq> tp = reqs(SeedData.RECIPES.get(2));
        List<S> p3 = Arrays.asList(new S(1, "pasta", 200, "g", null), new S(2, "tomatoes", 500, "g", null), new S(3, "onion", 1, "pcs", null),
                new S(4, "garlic", 2, "pcs", null), new S(5, "olive oil", 2, "tbsp", null), new S(6, "salt", 1, "tsp", null));
        t("mass covers count", RecipeMatcher.evaluate(3, tp, p3, true, today, 3).isReady());
        List<S> p4 = Arrays.asList(new S(1, "pasta", 200, "g", null), new S(2, "tomatoes", 500, "g", null), new S(3, "onion", 1, "cup", null),
                new S(4, "garlic", 2, "pcs", null), new S(5, "olive oil", 2, "tbsp", null), new S(6, "salt", 1, "tsp", null));
        MatchResult m4 = RecipeMatcher.evaluate(3, tp, p4, true, today, 3);
        t("mismatch not ready", !m4.isReady() && m4.firstProblem().status == MatchResult.Status.UNIT_MISMATCH);
        // expiry
        List<S> exp = new ArrayList<>(pantry); exp.replaceAll(s -> s.name.equals("Cheddar cheese") ? new S(s.id, s.name, s.q, s.unit, today.minusDays(1).toString()) : s);
        t("expired excluded", !RecipeMatcher.evaluate(11, reqs(SeedData.RECIPES.get(10)), exp, true, today, 3).isReady());
        t("expired allowed off", RecipeMatcher.evaluate(11, reqs(SeedData.RECIPES.get(10)), exp, false, today, 3).isReady());
        t("empty pantry", RecipeMatcher.evaluate(1, reqs(SeedData.RECIPES.get(0)), new ArrayList<S>(), true, today, 3).missingCount() == 4);
        t("bad unit safe", !RecipeMatcher.evaluate(1, reqs(SeedData.RECIPES.get(0)), Arrays.asList(new S(1, "egg", 3, "xx", null)), true, today, 3).isReady());
        // cooking
        Map<Long, Double> plan = RecipeMatcher.planCooking(reqs(SeedData.RECIPES.get(0)), pantry, true, today);
        t("cook eggs", plan.get(1L) == 3.0); t("cook butter", plan.get(4L) == 240.0); t("cook tomato", plan.get(2L) == 2.0);
        List<S> two = Arrays.asList(new S(1, "egg", 2, "pcs", "2026-09-25"), new S(2, "egg", 4, "pcs", "2026-09-21"));
        Map<Long, Double> pf = RecipeMatcher.planCooking(Arrays.asList(new Rq("eggs", 3, "pcs")), two, true, today);
        t("FEFO", pf.get(2L) == 1.0 && !pf.containsKey(1L));
        Map<Long, Double> del = RecipeMatcher.planCooking(Arrays.asList(new Rq("chicken", 300, "g")), Arrays.asList(new S(1, "chicken breast", 300, "g", null)), true, today);
        t("cook deletes", del.get(1L) == 0.0);
        t("days", RecipeMatcher.daysUntilExpiry("2026-09-23", today) == 3 && RecipeMatcher.daysUntilExpiry(null, today) == null);
        System.out.println(pass + " passed, " + fail + " failed");
        System.exit(fail == 0 ? 0 : 1);
    }
}

File: tests/web/engine.test.js
const E = require('../../web-preview/src/engine.js'); const R = require('../../web-preview/data/recipes.json'); const D = require('../../web-preview/data/demo_pantry.json');
let pass = 0, fail = 0; const t = (n, c) => { c ? pass++ : (fail++, console.log('FAIL', n)); };
const today = '2026-09-20';
const add = (d, n) => { const x = new Date(Date.UTC(2026, 8, 20 + n)); return x.toISOString().slice(0, 10); };
const o = { excludeExpired: true, today, soonDays: 3 };
// normalisation
t('tomatoes', E.canon('Tomatoes') === 'tomato'); t('potatoes', E.canon('potatoes') === 'potato');
t('eggs', E.canon('  EGGS ') === 'egg'); t('berries', E.canon('berries') === E.canon('berry'));
t('fresh adj', E.canon('Fresh large tomatoes') === 'tomato'); t('spaghetti', E.canon('Spaghetti') === 'pasta');
t('capsicum', E.canon('capsicum') === 'bell pepper'); t('yogurt', E.canon('Greek Yogurt') === 'yoghurt');
t('evoo', E.canon('extra virgin olive oil') === 'olive oil'); t('loaves', E.canon('loaves') === E.canon('loaf'));
t('cookie', E.canon('cookie') === E.canon('cookies')); t('hummus', E.canon('hummus') === 'hummus');
t('chilli flakes', E.canon('Chilli flakes') === E.canon('chili flake')); t('parens', E.canon('Tomatoes (cherry)') === 'tomato');
const pantry = D.map((d, i) => ({ id: i + 1, name: d.name, qty: d.qty, unit: d.unit, expiry: d.offset == null ? null : add(today, d.offset) }));
const res = R.map(r => E.evaluate(r, pantry, o));
const ready = res.filter(r => r.ready).map(r => r.recipeId), almost = res.filter(r => r.missingCount === 1).map(r => r.recipeId);
console.log('READY', ready.join(','), '| ALMOST', almost.join(','));
t('ready set', JSON.stringify(ready) === JSON.stringify([1, 2, 5, 6, 8, 11, 13]));
t('almost set', JSON.stringify(almost) === JSON.stringify([3, 4, 15]));
// strictness: 5 needed, 4 present -> not suggested
const only4 = pantry.filter(p => p.name !== 'Cheddar cheese'); const r4 = R.map(r => E.evaluate(r, only4, o));
t('remove cheddar drops 1,11,13', [1, 11, 13].every(id => !r4[id - 1].ready) && r4[1].ready);
// quantity: exactly enough passes, slightly less fails
const chick = R[4]; const mk = q => [{id:1,name:'chicken breast',qty:q,unit:'g',expiry:null},{id:2,name:'rice',qty:200,unit:'g',expiry:null},{id:3,name:'onion',qty:1,unit:'pcs',expiry:null},{id:4,name:'garlic',qty:2,unit:'pcs',expiry:null},{id:5,name:'soy sauce',qty:2,unit:'tbsp',expiry:null}];
t('exact qty ok', E.evaluate(chick, mk(300), o).ready); t('299 g fails', !E.evaluate(chick, mk(299), o).ready);
t('0.3 kg ok', E.evaluate(chick, mk(300).map((x, i) => i === 0 ? { ...x, qty: 0.3, unit: 'kg' } : x), o).ready);
// unit differences: soy sauce in ml/l/cup vs tbsp
const soy = u => mk(300).map((x, i) => i === 4 ? { ...x, qty: u[0], unit: u[1] } : x);
t('30 ml = 2 tbsp', E.evaluate(chick, soy([30, 'ml']), o).ready); t('29 ml fails', !E.evaluate(chick, soy([29, 'ml']), o).ready);
t('0.25 cup ok', E.evaluate(chick, soy([0.25, 'cup']), o).ready); t('1 l ok', E.evaluate(chick, soy([1, 'l']), o).ready);
// split batches are summed
const split = mk(300).map(x => x); split[0] = { ...split[0], qty: 150 }; split.push({ id: 9, name: 'Chicken', qty: 150, unit: 'g', expiry: null });
t('batches sum + alias chicken', E.evaluate(chick, split, o).ready);
// singular/plural in pantry vs recipe
const omel = R[0]; const p2 = [{id:1,name:'Egg',qty:3,unit:'pcs',expiry:null},{id:2,name:'Tomatoes',qty:1,unit:'pcs',expiry:null},{id:3,name:'Cheddar',qty:30,unit:'g',expiry:null},{id:4,name:'Butter',qty:10,unit:'g',expiry:null}];
t('plural/alias omelette', E.evaluate(omel, p2, o).ready);
// count vs mass: 500 g tomato covers 4 pcs (480 g)
const tp = R[2]; const p3 = [{id:1,name:'pasta',qty:200,unit:'g'},{id:2,name:'tomatoes',qty:500,unit:'g'},{id:3,name:'onion',qty:1,unit:'pcs'},{id:4,name:'garlic',qty:2,unit:'pcs'},{id:5,name:'olive oil',qty:2,unit:'tbsp'},{id:6,name:'salt',qty:1,unit:'tsp'}];
t('mass covers count', E.evaluate(tp, p3, o).ready);
// unknown conversion -> unit mismatch, not a crash
const p4 = p3.map(x => x.name === 'onion' ? { ...x, qty: 1, unit: 'cup' } : x);
const e4 = E.evaluate(tp, p4, o); t('mismatch handled', !e4.ready && e4.lines.find(l => l.key === 'onion').reason === 'short' || e4.lines.find(l => l.key === 'onion').reason === 'unit');
// expired excluded / included
const expired = pantry.map(p => p.name === 'Cheddar cheese' ? { ...p, expiry: add(today, -1) } : p);
t('expired excluded', !E.evaluate(R[10], expired, o).ready); t('expired allowed when off', E.evaluate(R[10], expired, { ...o, excludeExpired: false }).ready);
// empty pantry
t('empty pantry -> none ready', R.every(r => !E.evaluate(r, [], o).ready));
// zero-qty / bad unit ignored safely
t('bad unit safe', !E.evaluate(omel, [{id:1,name:'egg',qty:3,unit:'xx'}], o).ready);
// cooking deducts FEFO and deletes empty
const plan = E.planCooking(R[0], pantry, o); const eggs = pantry.find(p => p.name === 'Eggs'); 
t('cook: eggs 6->3', plan[eggs.id] === 3); t('cook: butter 250->240', plan[pantry.find(p => p.name === 'Butter').id] === 240);
t('cook: tomato 3->2', plan[pantry.find(p => p.name === 'Tomatoes').id] === 2);
const plan2 = E.planCooking(R[4], mk(300), o); t('cook: chicken deleted', plan2[1] === 0);
// FEFO across two batches
const fe = [{id:1,name:'egg',qty:2,unit:'pcs',expiry:'2026-09-25'},{id:2,name:'egg',qty:4,unit:'pcs',expiry:'2026-09-21'}];
const pf = E.planCooking({ingredients:[{name:'eggs',qty:3,unit:'pcs'}]}, fe, o); t('FEFO uses soonest first', pf[2] === 1 && pf[1] === undefined);
// cooking then re-evaluating: recipe no longer ready when pantry exhausted
const after = pantry.map(p => plan[p.id] === undefined ? p : { ...p, qty: plan[p.id] }).filter(p => p.qty > 0);
t('after cooking omelette still ready (plenty)', E.evaluate(R[0], after, o).ready);
console.log(pass + ' passed, ' + fail + ' failed'); process.exit(fail ? 1 : 0);
